torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+            if role_idx == 0 and replica_id == 0:
+                replica_role.env["TORCHX_RANK0_HOST"] = "localhost"
+
+            pod = role_to_pod(
+                name,
+                unique_app_id,
+                namespace,
+                replica_role,
+                service_account,
+                image_secret,
+            )
+            pod.metadata.labels.update(pod_labels(app, role_idx, role, replica_id))
+
+            genericitem: Dict[str, Any] = {
+                "replicas": 1,
+                "generictemplate": pod,
+            }
+            # TODO: retry support here
+            if role.max_retries > 0:
+                genericitem["maxRetry"] = role.max_retries
+                genericitem["policies"] = RETRY_POLICIES[role.retry_policy]
+                msg = f"""
+Role {role.name} configured with restarts: {role.max_retries}.
+                """
+                warnings.warn(msg)
+
+            genericitems.append(genericitem)
+
+    """
+    Create Service:
+    The selector will have the key 'appwrapper.mcad.ibm.com', and the value will be 
+    the appwrapper name
+    """
+
+    service_port = get_port_for_service(app)
+
+    if not (0 < int(service_port) <= 65535):
+        msg = """Warning: rdzv_endpoint set to invalid port number. Value must be between 1-65535, with torchx default = 29500. Try updating --rdzv_port argument to dist.ddp. Setting port to default = 29500"""
+        service_port = "29500"
+        warnings.warn(msg)
+
+    svc_obj = mcad_svc(unique_app_id, namespace, service_port)
+
+    genericitem_svc: Dict[str, Any] = {
+        "replicas": 1,
+        "generictemplate": svc_obj,
+    }
+    genericitems.append(genericitem_svc)
+
+    job_spec: Dict[str, Any] = {
+        "resources": {
+            "GenericItems": genericitems,
+        },
+    }
+
+    if priority is not None:
+        job_spec["priority"] = priority
+
+    resource: Dict[str, object] = {
+        "apiVersion": "mcad.ibm.com/v1beta1",
+        "kind": "AppWrapper",
+        "metadata": {"name": unique_app_id},
+        "spec": job_spec,
+    }
+    return resource
+
+
+# Helper functions for MCAD generic items information -> TorchX Role
+def get_role_information(genericItems: Iterable[Dict[str, Any]]) -> Dict[str, Any]:
+    # Store unique role information
+    roles = {}
+
+    # nested dictionary keys
+    # meta data information
+    gt_key = "generictemplate"
+    metadata_key = "metadata"
+    label_key = "labels"
+    role_key = "torchx.pytorch.org/role-name"
+
+    # containers information
+    spec_key = "spec"
+    container_key = "containers"
+    image_key = "image"
+    args_key = "command"
+    env_key = "env"
+    resource_key = "resources"
+    ports_key = "ports"
+    mounts_key = "volumeMounts"
+
+    # resource keys
+    cpu_key = "cpu"
+    gpu_key = "gpu"
+    request_key = "requests"
+    mem_key = "memory"
+
+    for genericItem in genericItems:
+        if gt_key in genericItem.keys():
```

Comment:
this is really nested, can we flatten a lot of these out with continues?

i.e.

```
if gt_key not in genericItem.keys():
   continue
gt_result = generic_item[gt_key]
...
```

Reply:
Restructured in commit: 2c4a0c91493960d11a25068ab7729ac5eaf25177